src/v/iceberg/metadata_io.h

```diff
@@ -0,0 +1,187 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "base/outcome.h"
+#include "base/vlog.h"
+#include "bytes/iobuf.h"
+#include "cloud_io/remote.h"
+#include "cloud_storage_clients/types.h"
+#include "iceberg/logger.h"
+#include "ssx/future-util.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/coroutine/as_future.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+namespace iceberg {
+
+using namespace std::chrono_literals;
+class metadata_io {
+public:
+    enum class errc {
+        // The call has timed out. Callers should retry, if allowed.
+        timedout = 1,
+
+        // The call failed for some reason that doesn't appear to be retriable.
+        failed,
+
+        // The call failed because the system is shutting down.
+        shutting_down,
+    };
+    metadata_io(cloud_io::remote& io, cloud_storage_clients::bucket_name b)
+      : io_(io)
+      , bucket_(std::move(b)) {}
+    ~metadata_io() = default;
+
+protected:
+    template<typename T>
+    ss::future<checked<T, errc>> download_object(
+      const std::filesystem::path& path,
+      std::string_view display_str,
+      ss::noncopyable_function<T(iobuf)> parse) {
```

Comment:
consider taking parameters by value to coroutines that aren't in a hotpath. for example, display_str is accessed after a co_await, so it's necessary to consider if its lifetime extension.